### ISSUE TITLE
Synchronisation primitives for async code

### ISSUE THREAD

**Issue by ashfordneil** [rustasync/team#33]
In synchronous rust, we have great things like mutexes and channels and barriers. In asynchronous rust, we have async-channels that are pretty cool, but we don't have the rest. At this point, it feels like the only current option is to use the synchronous synchronization primitives in your asynchronous code, but this has issues, mainly tied around the fact that the `std::sync` primitives are scheduled by the operating system and futures are scheduled by the executor. Some issues that I can think of are

- you can't `await!` a synchronous primitive being ready to unblock with the way they're currently implemented
- if you `await!` something else while holding a lock on a mutex a whole bunch of things may go wrong deadlock-wise

# My suggestion
We have async channels in the futures crate that are basically mirrors of the sync channels that instead use our own scheduling / waking logic. I suggest making something like `futures::sync::{Mutex, Barrier, ...}` to do a similar thing. Thanks to the change to explicit contexts and wakers with futures 0.2/0.3, the code behind the old `BiLock` could be refactored to be a lot more safe, and extended to form more general (and publicly exposed) asynchronous synchronization primitives.

# Alternatives I'm ware of
## keep the normal synchronous primitives, but do compiler magic to double check the interaction between them and async code (or futures in general) isn't going to cause issues
Not my idea, so I may not have given it the best justice in this explanation.
If I understand it correctly, I'd have some worries that such checks would be really hard without a lot of false negatives or false positives in terms of what is considered "not a good idea in `async`".

## just let people model everything as channels
Seems to be working pretty fine for go, and if I understand correctly erlang. However, there's a reason we didn't go with this model for the synchronous synchronization primitives. I worry that there are situations where you'd suffer potentially unacceptable performance losses modelling your synchronization purely through channels instead of using less abstract synchronisation primitives.

**Comment by polachok** [comment]
Go has mutexes and it is often recommended to use them instead of channels.

**Comment by ashfordneil** [comment]
Also I forgot to mention - I would 100% be keen to put the hours in myself to write such asynchronous synchronization primitives. Idk if that helps, but it seemed worth mentioning.

**Comment by jsgf** [comment]
I guess there's two things to call out here specifically:
1. Sync Mutex and similar operations are a very bad fit with async code. This will get worse with async/await because it'll be easy to accidentally combine them. We should at least have warnings for this (clippy), or outright errors (typesystem).
2. The existing set of async-compatible Mutex operators is more limited. In async/await code, they'll be more useful because the overall code style will be more similar to sync code, and it would be nice to have drop-in analogs for the sync variants (which would help mitigate 1).

For my part, I've found very little use for a Mutex-like thing in combinator async code. Whenever I've reached for one I've always solved the problem more cleanly some other way.

But I think with async/await code they will be more useful both as a transitional tool and overall.

I'm not very keen about magically making sync Mutex operations async compatible. I think there'll be enough subtle differences that it will just complicate things (ie, the apparently simple model has tons of hidden state which makes it actually much more complex).

**Comment by erickt** [comment]
I also have been thinking about this over the past week. It's pretty easy to see how we can hit a deadlock with a single-threaded executor:

```
async fn do_foo(thing: Arc<Mutex<Thing>>, ...) -> ... {
    async {
        // 1. first we grab the lock
        let thing = self.thing.lock().unwrap();

        // 2. yield to the scheduler
        let signed = await!(
            repo.fetch_metadata(...)
        );

        // 4. we'll never resume here because we're blocked waiting on the lock.

        ...
    }
}

async fn do_bar(thing: Arc<Mutex<Thing>>, ...) -> ... {
    async {
        // 3. we try to grab the lock and hang.
        let thing = self.thing.lock().unwrap();

        ...
    }
```

Multi-threaded executors with `Send`-able futures however should  get a compile error, because `MutexGuard` is `!Send`, and therefore cannot be captured in the async/await state.

We could go even further though, and prevent this with a new inherent trait, which I'm calling `?Yield`, which says if the type is safe to be alive across a yield point. We then could then opt-out `MutexGuard` from implementing this trait. I think this is somewhat analogous to C# not implementing `lock { ... }` inside [async blocks](https://stackoverflow.com/questions/7612602/why-cant-i-use-the-await-operator-within-the-body-of-a-lock-statement#7612714).

Would this be worthwhile? To answer this, the questions I have are:

* Are there any other kinds of things that are dangerous to keep across a yield point?
* How much do we care about single-threaded executors, since I imagine making guards `!Send` would catch many problems. However, I did see that the [lock_api](https://docs.rs/lock_api/0.1.3/lock_api/struct.GuardSend.html) crate (which is used by the [parking_lot](https://github.com/Amanieu/parking_lot) crate) has the ability to define sendable mutex guards.
* Are there any justifiable reasons to actually want to hold a type like `MutexGuard` across a yield point? I'm guessing there's probably none, but it's worth thinking through.
* It's a lot easier to solve this on the library-side, by defining async-compatible primitives. [C# does this with SemaphoreSlim](https://msdn.microsoft.com/en-us/library/hh462723.aspx). Is it sufficient to solve this with education, and just tell people "use this library, not that library" with async/await functions?

**Comment by tynril** [comment]
This is probably a very naive proposal. Would it be possible to implement some kind of `Yield` trait, similar to `Drop`, such as:

```
pub trait Yield {
    /// Called on all local values in an `async` block upon reaching a yield point.
    fn yield(&mut self);

    /// Called on all local values when resuming an `async` block after a yield point.
    fn resume(&mut self);
}
```

So that we could implement this trait on `MutexGuard` in such as way that the lock is released in `yield` and re-acquired in `resume`?

**Comment by ashfordneil** [comment]
@erickt - your earlier comment
> Are there any justifiable reasons to actually want to hold a type like MutexGuard across a yield point? I'm guessing there's probably none, but it's worth thinking through

Wouldn't you want to do this any time you had atomic / critical regions of code that needed to do IO or something else that involves waiting / blocking? Naively, my mind turns to updating configuration settings over a network in a running process, but I'm sure there would be other uses.

Re: your other dot points, I agree with you on the "do it at a library level" point, and would need to research more to discuss your other points.

**Comment by Pzixel** [comment]
I faced the issue in my real app. It's a web service that saves users images.

So I'm writing something like:

```rust
async fn handle_request(db: Arc<Mutex<ImageDatabase>>) {
   let file = await!(get_file());
   let db = db_mutex.lock().unwrap();
   let does_contain_file = db.contains_file(file);
   if !does_contain_file {
      db.insert(file);
   }
   await!(send_request(does_contain_file))
}
```

And now i'm a bit lost, i'm not sure how should I proceed.

**Comment by Nemo157** [comment]
The latest version of `futures-preview` (`0.3.0-alpha.10`+) includes a futures-aware mutex: [`futures::lock::Mutex`](https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.10/futures/lock/struct.Mutex.html).

**Comment by najamelan** [comment]
@Nemo157 cool, thanks for pointing that out. It would be nice if someone who understands that mutex wrote up a few paragraphs of documentation. Is it meant for multithreaded scenarios or for single threaded executors? 

Do I get it right that this is the principle advantage of an actor model, eg. you get the guarantee that an actor will handle one message at a time, so it's internal state will remain consistent.

I wonder if the price of the messaging would be more or less than the price of using synchronization? Apart from the conveniently simple mental model compared to having to think about the danger of every await statement.

It seems to me where rust holds your hand  for multithreading, in the case of async, we are back to C++ kind of responsibility of the developer. Is this async fn re-entrant? Could it be called from several tasks during yields? It's very easy to let errors slip in.
